### PR TITLE
use RepeatTestAction rather than id strings to track the most recent …

### LINKED ISSUE #26577: fattest.simplicity tracks the active feature request as strings and this causes bad code
fattest.simplicity stores a stack of strings, each string being the ID of a RepeatTestAction, and the stack tracks which action is being run.

Because it stores only strings we have some bad code doing if/else if on every possibility to find which class it needs to invoke. This will only get longer as we make more repeat actions. This also means that the code is not easily extensible if we need to add another unique action for future runtimes like we did when we introduced the transformer.

To fix this, simply put the actual RepeatTestAction in the stack and then invoke on the current action where needed.

### PR DESCRIPTION
**Pull request by benjamin-confino**
…action

fixes #26577

### CODE REVIEW

File: dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEE10Action.java [hunk 1]
```diff
@@ -222,7 +222,8 @@ public JakartaEE10Action withWiden() {
      * @param newAppPath                The application path of the transformed file (or <code>null<code>)
      * @param transformationRulesAppend The map with the additional transformation rules to add
      */
-    static void transformApplication(Path appPath, Path newAppPath, Map<String, String> transformationRulesAppend) {
+    @Override
+    public void transformApplication(Path appPath, Path newAppPath, Map<String, String> transformationRulesAppend) {
```
Comment (Azquelt) [inline dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEE10Action.java]: Why is this change wanted? It doesn't seem related to the description of the PR

File: dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java [hunk 2]
```diff
@@ -255,13 +255,13 @@ public static void transformApp(Path appPath, EEVersion eeVersion) {
     public static void transformApp(Path appPath, Path newAppPath, EEVersion eeVersion) {
         switch (eeVersion) {
             case EE9:
-                JakartaEE9Action.transformApplication(appPath, newAppPath, null);
+                new JakartaEE9Action().transformApplication(appPath, newAppPath, null);
                 break;
```
Comment (Azquelt) [inline dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java]: Kinda gross. You need to create an instance just to transform the app.

File: dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java [hunk 3]
```diff
@@ -276,13 +276,7 @@ public static void transformApp(Path appPath, Path newAppPath, EEVersion eeVersi
...
      */
     public static void transformApp(Path appPath, Path newAppPath, Map<String, String> transformationRulesAppend) {
-        if (RepeatTestFilter.isRepeatActionActive(EE9_ACTION_ID)) {
-            JakartaEE9Action.transformApplication(appPath, newAppPath, transformationRulesAppend);
-        } else if (RepeatTestFilter.isRepeatActionActive(EE10_ACTION_ID)) {
-            JakartaEE10Action.transformApplication(appPath, newAppPath, transformationRulesAppend);
-        } else if (RepeatTestFilter.isRepeatActionActive(EE11_ACTION_ID)) {
-            JakartaEE11Action.transformApplication(appPath, newAppPath, transformationRulesAppend);
-        }
+        RepeatTestFilter.getMostRecentRepeatAction().transformApplication(appPath, newAppPath, transformationRulesAppend);
```
Comment (tevans78) [inline dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java]: This logic has been changed ... it now only checks only the top action on the stack rather than all of them. This change may break some tests.

File: dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java (hunk 2 above)
Comment (tevans78) [reply dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java]: We could have both a static method and a non-static one? Although really the transformation code should be separated slightly from the Action (a different issue).

File: dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEE10Action.java (hunk 1 above)
Comment (jhanders34) [reply dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEE10Action.java]: Yeah I purposely don't have this be public.  I want people to use the JakartaEEAction class to do transformation calls.

File: dev/fattest.simplicity/src/componenttest/rules/repeater/RepeatTestAction.java [hunk 4]
```diff
@@ -35,4 +38,6 @@ public interface RepeatTestAction {
      * If clean up is needed to undo the setup changes after running the action then override this method.
      */
     default public void cleanup() throws Exception {}
+
+    default public void transformApplication(Path appPath, Path newAppPath, Map<String, String> transformationRulesAppend) {}
```
Comment (jhanders34) [inline dev/fattest.simplicity/src/componenttest/rules/repeater/RepeatTestAction.java]: Yeah I don't want this.

Comment (jhanders34) [reply dev/fattest.simplicity/src/componenttest/rules/repeater/RepeatTestAction.java]: I expect people to use the following method on JakartaEEAction in order to transform to a specific EE level:

```
    public static void transformApp(Path appPath, Path newAppPath, EEVersion eeVersion)
```

File: dev/fattest.simplicity/src/componenttest/custom/junit/runner/RepeatTestFilter.java [hunk 5]
```diff
@@ -22,6 +22,7 @@
 import org.junit.runners.model.FrameworkMethod;
 
 import componenttest.annotation.SkipForRepeat;
+import componenttest.rules.repeater.RepeatTestAction;;
```
Comment (jhanders34) [inline dev/fattest.simplicity/src/componenttest/custom/junit/runner/RepeatTestFilter.java]:
```suggestion
import componenttest.rules.repeater.RepeatTestAction;
```

File: dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java (hunk 3 above)
Comment (jhanders34) [reply dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java]: Yeah I don't like this at all.  I don't understand why this change is necessary.

Comment (benjamin-confino) [reply dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java]: Actually the logic hasn't changed. Due to the bizarre code for repeat tests the stack only ever has one repeat test action on it at any given time. So its already only checking the top action. I think if the stack was actually used as a stack the code I replaced would cause a lot of bugs.

As for why its necessary. Currently if you do `FeatureReplacementAction.EE10_FEATURES().withID("my_id")` apps will not be transformed because transformation runs based on string comparisons from the IDs. This is very unintuitive behaviour and this PR was created in response to someone tripping over it. With this new code transformation is a property of the class itself which leads to far more logical behaviour.

File: dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEE10Action.java (hunk 1 above)
Comment (benjamin-confino) [reply dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEE10Action.java]: Currently the code is using string comparison to determine if an app should transform. I'm changing that to a more reliable system where it gets the current action as an object and invokes its transform method. 

That required putting the transform method on the interface, and interface methods can only be public. Though given Jarid's comments I'll have to find a different way.

File: dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java [hunk 6]
```diff
@@ -276,12 +276,10 @@ public static void transformApp(Path appPath, Path newAppPath, EEVersion eeVersi
...
     public static void transformApp(Path appPath, Path newAppPath, Map<String, String> transformationRulesAppend) {
-        if (RepeatTestFilter.isRepeatActionActive(EE9_ACTION_ID)) {
-            JakartaEE9Action.transformApplication(appPath, newAppPath, transformationRulesAppend);
-        } else if (RepeatTestFilter.isRepeatActionActive(EE10_ACTION_ID)) {
-            JakartaEE10Action.transformApplication(appPath, newAppPath, transformationRulesAppend);
-        } else if (RepeatTestFilter.isRepeatActionActive(EE11_ACTION_ID)) {
-            JakartaEE11Action.transformApplication(appPath, newAppPath, transformationRulesAppend);
+        RepeatTestAction action = RepeatTestFilter.getMostRecentRepeatAction();
+
+        if (action instanceof FeatureReplacementAction) {
```
Comment (jhanders34) [inline dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java]: This should be an instanceof change of JakartaEEAction.  The point of having this common abstract class is not have have the logic in FeatureReplacementAction, but instead of have it in this class.

Comment (jhanders34) [reply dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java]: Can just do instanceof changes for JakartaEEAction9, 10 and 11 and just call their static method as well instead of having a new non static method.

File: dev/fattest.simplicity/src/componenttest/rules/repeater/FeatureReplacementAction.java [hunk 7]
```diff
@@ -899,4 +899,8 @@ public String getID() {
             return getClass().getSimpleName() + "  REMOVE " + removeFeatures + "  ADD " + addFeatures;
         }
     }
+
+    void transformApplication(Path appPath, Path newAppPath, Map<String, String> transformationRulesAppend) {
+        // no-op
+    }
```
Comment (jhanders34) [inline dev/fattest.simplicity/src/componenttest/rules/repeater/FeatureReplacementAction.java]:
```suggestion
```

File: dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java [hunk 8]
```diff
@@ -299,12 +297,10 @@ public static void transformApp(Path appPath, Path newAppPath, Map<String, Strin
...
-        } else if (RepeatTestFilter.isRepeatActionActive(EE10_ACTION_ID)) {
-            JakartaEE10Action.transformApplication(appPath, newAppPath, null);
-        } else if (RepeatTestFilter.isRepeatActionActive(EE11_ACTION_ID)) {
-            JakartaEE11Action.transformApplication(appPath, newAppPath, null);
+        RepeatTestAction action = RepeatTestFilter.getMostRecentRepeatAction();
+
+        if (action instanceof FeatureReplacementAction) {
+            ((FeatureReplacementAction) action).transformApplication(appPath, newAppPath, null);
         }
     }
```
Comment (jhanders34) [inline dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java:317]:
```suggestion
abstract void transformApplication(Path appPath, Path newAppPath, Map<String, String> transformationRulesAppend);
```

File: dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java [hunk 9]
```diff
@@ -299,12 +297,10 @@ public static void transformApp(Path appPath, Path newAppPath, Map<String, Strin
...
-        if (RepeatTestFilter.isRepeatActionActive(EE9_ACTION_ID)) {
-            JakartaEE9Action.transformApplication(appPath, newAppPath, null);
-        } else if (RepeatTestFilter.isRepeatActionActive(EE10_ACTION_ID)) {
-            JakartaEE10Action.transformApplication(appPath, newAppPath, null);
-        } else if (RepeatTestFilter.isRepeatActionActive(EE11_ACTION_ID)) {
-            JakartaEE11Action.transformApplication(appPath, newAppPath, null);
+        RepeatTestAction action = RepeatTestFilter.getMostRecentRepeatAction();
+
+        if (action instanceof FeatureReplacementAction) {
+            ((FeatureReplacementAction) action).transformApplication(appPath, newAppPath, null);
```
Comment (jhanders34) [inline dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java]:
```suggestion
        if (action instanceof JakartaEEAction) {
            ((JakartaEEAction) action).transformApplication(appPath, newAppPath, null);
```

File: dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java [hunk 10]
```diff
@@ -276,12 +276,10 @@ public static void transformApp(Path appPath, Path newAppPath, EEVersion eeVersi
...
-        if (RepeatTestFilter.isRepeatActionActive(EE9_ACTION_ID)) {
-            JakartaEE9Action.transformApplication(appPath, newAppPath, transformationRulesAppend);
-        } else if (RepeatTestFilter.isRepeatActionActive(EE10_ACTION_ID)) {
-            JakartaEE10Action.transformApplication(appPath, newAppPath, transformationRulesAppend);
-        } else if (RepeatTestFilter.isRepeatActionActive(EE11_ACTION_ID)) {
-            JakartaEE11Action.transformApplication(appPath, newAppPath, transformationRulesAppend);
+        RepeatTestAction action = RepeatTestFilter.getMostRecentRepeatAction();
+
+        if (action instanceof FeatureReplacementAction) {
+            ((FeatureReplacementAction) action).transformApplication(appPath, newAppPath, transformationRulesAppend);
```
Comment (jhanders34) [inline dev/fattest.simplicity/src/componenttest/rules/repeater/JakartaEEAction.java]:
```suggestion
        if (action instanceof JakartaEEAction) {
            ((JakartaEEAction) action).transformApplication(appPath, newAppPath, transformationRulesAppend);
```

**Review by jhanders34 — CHANGES_REQUESTED** [review]
This will make it the way I think it should be.

**Review by jhanders34 — DISMISSED** [review]

**Review by jhanders34 — DISMISSED** [review]

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
#build

**Review by jhanders34 — APPROVED** [review]

**Comment by jhanders34** [comment]
#libby

**Comment by benjamin-confino** [comment]
#build

**Comment by benjamin-confino** [comment]
one final build since the last one was a week ago.

File: dev/io.openliberty.cdi.core_fat/fat/src/io/openliberty/cdi/internal/core/RepeatRule.java [hunk 11]
```diff
@@ -38,16 +39,16 @@
  */
 public class RepeatRule<T> implements TestRule {
 
-    private T[] repeats;
+    private RepeatTestAction[] repeats;
 
-    private T currentRepeat;
+    private RepeatTestAction currentRepeat;
```
Comment (Azquelt) [inline dev/io.openliberty.cdi.core_fat/fat/src/io/openliberty/cdi/internal/core/RepeatRule.java:44]: Every use of `T` has been removed. The type parameter should have been removed as well and the Javadoc updated.
